enter.c

```diff
@@ -110,24 +102,51 @@ int enter(struct entry_settings *opts)
 	}
 	root = resolved_root;
 
-	int unshareflags = opts_to_unshareflags(opts);
-
-	struct outer_helper outer_helper;
-
-	if (unshareflags & BST_CLONE_NEWUSER) {
-		outer_helper = outer_helper_spawn();
-	}
-
 	int timens_offsets = -1;
-	if (unshareflags & BST_CLONE_NEWTIME) {
+	if (opts->shares[SHARE_TIME] != share_with_parent) {
 		timens_offsets = open("/proc/self/timens_offsets", O_WRONLY);
 		if (timens_offsets == -1) {
 			if (errno != ENOENT) {
 				err(1, "open(\"/proc/self/timens_offsets\")");
 			}
-			/* The kernel evidently don't support time namespaces yet. No need
-			   to try below. */
-			unshareflags &= ~BST_CLONE_NEWTIME;
+			/* The kernel evidently doesn't support time namespaces yet.
+			   Don't try to open the time namespace file with --share-all=<dir>,
+			   or try to unshare or setns the time namespace below. */
+			opts->shares[SHARE_TIME] = share_with_parent;
+		}
+	}
+	
+	int nsactions[MAX_SHARES];
+	opts_to_nsactions(opts, nsactions);
+
+	struct outer_helper outer_helper;
+	outer_helper.persist = opts->persist;
+	outer_helper.unshare_user = nsactions[SHARE_USER] == NSACTION_UNSHARE;
+	outer_helper_spawn(&outer_helper);
+
+	/* While we still have privileges, setns into all relevant
```

Comment:
This really should be done unprivileged. Did you have time to look at my PoC?

Reply:
done, thanks, sorry for dropping this